File: sdk/src/main/java/com/robotemi/sdk/listeners/OnBeWithMeStatusChangedListener.java
package com.robotemi.sdk.listeners;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import androidx.annotation.StringDef;

public interface OnBeWithMeStatusChangedListener {

    String ABORT = "abort";

    String CALCULATING = "calculating";

    String LOCK = "lock";

    String SEARCH = "search";

    String START = "start";

    String TRACK = "track";

    String OBSTACLE_DETECTED = "obstacle detected";

    /**
     * Listen for status changes during 'beWithMe'.
     * <p>
     * Available statuses:
     * <ul>
     * <li>{@link OnBeWithMeStatusChangedListener#ABORT}</li>
     * <li>{@link OnBeWithMeStatusChangedListener#CALCULATING}</li>
     * <li>{@link OnBeWithMeStatusChangedListener#LOCK}</li>
     * <li>{@link OnBeWithMeStatusChangedListener#SEARCH}</li>
     * <li>{@link OnBeWithMeStatusChangedListener#START}</li>
     * <li>{@link OnBeWithMeStatusChangedListener#TRACK}</li>
     * <li>{@link OnBeWithMeStatusChangedListener#OBSTACLE_DETECTED}</li>
     * </ul>
     *
     * @param status Current status.
     */
    void onBeWithMeStatusChanged(@BeWithMeStatus String status);

    @StringDef({ABORT, CALCULATING, LOCK, SEARCH, START, TRACK, OBSTACLE_DETECTED})
    @Retention(RetentionPolicy.SOURCE)
    @interface BeWithMeStatus {
    }
}

File: sdk/src/main/java/com/robotemi/sdk/listeners/OnBeWithMeStatusChangedListener.kt
package com.robotemi.sdk.listeners

import androidx.annotation.StringDef

interface OnBeWithMeStatusChangedListener {

    /**
     * Listen for status changes during 'beWithMe'.
     *
     *
     * Available statuses:
     *
     *  * [OnBeWithMeStatusChangedListener.ABORT]
     *  * [OnBeWithMeStatusChangedListener.CALCULATING]
     *  * [OnBeWithMeStatusChangedListener.SEARCH]
     *  * [OnBeWithMeStatusChangedListener.START]
     *  * [OnBeWithMeStatusChangedListener.TRACK]
     *  * [OnBeWithMeStatusChangedListener.OBSTACLE_DETECTED]
     *
     *
     * @param status Current status.
     */
    fun onBeWithMeStatusChanged(@BeWithMeStatus status: String)

    @StringDef(ABORT, CALCULATING, SEARCH, START, TRACK, OBSTACLE_DETECTED)
    @kotlin.annotation.Retention(AnnotationRetention.SOURCE)
    annotation class BeWithMeStatus

    companion object {

        const val ABORT = "abort"

        const val CALCULATING = "calculating"

        const val SEARCH = "search"

        const val START = "start"

        const val TRACK = "track"

        const val OBSTACLE_DETECTED = "obstacle detected"
    }
}
